docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+- Feature Name: Partition Hibernation
+- Status: draft
+- Start Date: 2022-04-08
+- Authors: John Spray
+- Issue: (one or more # from the issue tracker)
+
+# Executive Summary
+
+Avoid idle partitions using system resources.
+
+## What is being proposed
+
+A new state for partitions, where they exist in the topics table
+but do not have active raft groups.
+
+## Why (short reason)
+
+To enable higher scale when a significant proportion of topics
+are idle.
+
+## How (short plan)
+
+Topics will be automatically placed into the hibernate state when
+they have been idle for more than a configurable threshold.
+
+## Impact
+
+Reduced cost of delivering free/cheap public-facing services, where
+users are likely to create topics for occasional or one-off use.
+
+# Motivation
+
+## Why are we doing this?
+
+Partitions have some runtime resource overhead, even if they aren't being
+used, and haven't been used for some time.
+
+To support public-facing Kafka services on which many users might create
+and then abandon idle topics, we could implement a "hibernate" mode
+for partitions that haven't been used for a long time.
+
+## What use cases does it support?
+
+Implementing a low-cost/no-cost public service, in which developers are likely
+to create unused topics with which to experiment.
+
+## What is the expected outcome?
+
+Systems will run smoothly with a higher number of partitions than
+they could before, as long as some of those partitions are idle.
+
+# Guide-level explanation
+
+## Introducing new named concepts.
+
+`Hibernated topic`: a topic whose partitions do not have active raft
+groups running in the cluster.  It cannot serve user I/O (aside from
+metadata requests) until it is woken up.
+
+`Waking a topic`: a control plane operation to change the status of
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
```

Comment:
the sentence

> There is a "hibernating" intermediate state where members of the
> partitions' raft groups are responsible for refusing writes, and

sounds like the partition is not idle in which case it should be hibernated?

Reply:
It's an edge case, where the partition has been idle for long enough to hibernate, but happens to receive an I/O just as it's entering hibernation.   The thinking behind write-blocking is that it's simpler to have a "point of no return" than to have a mechanism for aborting hibernation when this happens, but I didn't go all the way down the rabbit hole of details, maybe aborting is viable.

Reply:
> simpler to have a "point of no return" than to have a mechanism for aborting hibernation when this happens

got it. this is implementation level discussion it seems related to synchronization that should be straight forward to address.